### PR TITLE
setup: Add fish support

### PR DESCRIPTION
**Pull request by superm1**
This reverts some of 8ea478a ("fish: Remove fish support") but fixes the unit tests and adds coverage for all remaining commands

Fixes: https://github.com/kworkflow/kworkflow/issues/1165

### CODE REVIEW

**Comment by rodrigosiqueira** [comment]
I was reviewing your change before you closed it; you can see my partial review at:

https://github.com/kworkflow/kworkflow/pull/1166/files

**Comment by superm1** [comment]
> I was reviewing your change before you closed it; you can see my partial review at:
> 
> https://github.com/kworkflow/kworkflow/pull/1166/files

Thanks. I've incorporated the additional feedback.

File: tests/unit/fish_completion_test.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                      explore e pomodoro p report r device backup debug
+                      send-patch env patch-hub drm vm clear-cache codestyle
+                      c version v man help h"
+  local sorted_kw_options="$(echo $kw_options | awk '{gsub(" ", "\n"); print $0}' | sort)"
+  local sorted_kw_fish_completions="$(get_sorted_completions 'kw')"
+
+  if [[ "$sorted_kw_fish_completions" != "$sorted_kw_options" ]]; then
+    local difference="$(diff <(echo "$sorted_kw_options") <(echo "$sorted_kw_fish_completions"))"
+    fail "Fish suggestions and kw commands are not the same:\n$difference"
+  fi
```
Comment (rodrigosiqueira) [inline tests/unit/fish_completion_test.sh:32]: You can replace all of this logic with something like this:

` assert_equals_helper 'Autocomplete is different for fish' "$LINENO" "${sorted_kw_fish_completions}" "$sorted_kw_options"

File: src/kw.fish
Comment (rodrigosiqueira) [inline src/kw.fish:1]: This part, I'll rely on you since I don't know enough about fish :)

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hey Mario,

Thanks a lot for your updated PR. I added some other comments. Thanks

File: tests/unit/fish_completion_test.sh [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+function get_sorted_completions()
+{
+  local env_path=".:/bin"
+  local env_fish_complete_path=""
+  local completion="./src/kw.fish"
+  local cmd=$1
+  local fish_cmd="set PATH $env_path;
+                    set fish_complete_path $env_fish_complete_path;
+                    source $completion;
+                    complete -C'$1 '"
```
[the remark this answers is not in the document]
Comment (superm1) [reply tests/unit/fish_completion_test.sh:14]: For whitespace?  Or for something else?

File: tests/unit/fish_completion_test.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+function test_kw_completion_for_fish()
+{
+  local kw_options="init build b deploy d bd diff df ssh s self-update u
+                      maintainers m kernel-config-manager k config g remote
+                      explore e pomodoro p report r device backup debug
+                      send-patch env patch-hub drm vm clear-cache codestyle
+                      c version v man help h"
+  local sorted_kw_options="$(echo $kw_options | awk '{gsub(" ", "\n"); print $0}' | sort)"
```
[the remark this answers is not in the document]
Comment (superm1) [reply tests/unit/fish_completion_test.sh:26]: How pedantic are you on `printf` vs `echo`?  AFAICT this is a case of `echo foo` vs `printf '%s%s%s%s%s%s'` with a gazillion strings.

File: tests/unit/fish_completion_test.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply tests/unit/fish_completion_test.sh:14]: Sorry for my vague comment; I meant the variable concatenation.

[the remark this answers is not in the document]
Comment (superm1) [reply tests/unit/fish_completion_test.sh:14]: Can you suggest what needs to be changed specifically?  I'm having a hard time following.

This is invoked in the `fish` shell not the `bash` shell.

[the remark this answers is not in the document]
Comment (rodrigosiqueira) [reply tests/unit/fish_completion_test.sh:14]: I meant this:

```
  local fish_cmd="set PATH ${env_path};
                    set fish_complete_path ${env_fish_complete_path};
                    source $completion;
                    complete -C'${1} '"
```

This is still a bash variable, and all the internal variables will be expanded before fish take control (`fish -c "$fish_cmd" | awk '{print $1}' | sort`).

**Comment by rodrigosiqueira** [comment]
Hi @superm1 , thanks a lot for your PR.

Your change was merged into the unstable branch.

**Pull request closed without merging**